[v4] Starting v4 generated project works again (@W-19169283@)

vmarta opened the pull request:
When we generated a project and running it, we saw this error `TypeError: Cannot read properties of undefined (reading '_config')` in the browser. The error came up only with _generated_ projects.

So this PR fixes the problem by making sure that all static assets (like the `.hbs` files) found in the generator are in sync and up to date.

## How to test drive the PR

Generate a project and run it:

1. From the monorepo root, run `node packages/pwa-kit-create-app/scripts/create-mobify-app-dev.js --outputDir /tmp/generated-project-v4/ --preset chakra-storefront-demo`
2. Switch to that generated project folder
3. `npm start`
4. Verify that the project runs as expected

vmarta committed:
Sync the AppConfig hbs file

vmarta committed:
We no longer use the manifest file

vmarta committed:
Sync the ssr.js.hbs too

cc-prodsec commented:
### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/5e8526c2-62c9-45e8-be11-14ce1a123ab9?refs=d13ea69893ac585c98d276b83ee9a94afbea37da&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/5e8526c2-62c9-45e8-be11-14ce1a123ab9/license?refs=d13ea69893ac585c98d276b83ee9a94afbea37da&source=prChecksComment)

vmarta committed:
Revert "We no longer use the manifest file"

This reverts commit 45ec28bf89e63647be3fa32c99efef9f959ab70c.

vmarta committed:
Remove comments

vmarta committed:
Empty commit

vmarta commented on packages/pwa-kit-create-app/assets/templates/chakra-storefront/src/components/_app-config/index.jsx.hbs:
```diff
@@ -44,33 +40,55 @@ const AppConfig = ({children, locals = {}}) => {
         >
             <MultiSiteProvider site={locals.site} locale={locals.locale} buildUrl={locals.buildUrl}>
-                <ChakraProvider theme={theme}>{children}</ChakraProvider>
+                <ChakraProvider value={theme}>{children}</ChakraProvider>
```
In Chakra v3, `value` is the correct prop to use here. This change is what fixes the `_config` error. 

But we needed to make sure the entire file (and other files too) is up to date.

vmarta committed:
Merge branch 'next' into vm/fix-v4-generator

vmarta committed:
Bring back the manifest.json file

vmarta committed:
Fix warning about invalid nested Helmet

Helmet is not needed here, because it's already provided by the containing Seo component.

vmarta commented on packages/pwa-kit-create-app/assets/templates/chakra-storefront/src/ssr.js.hbs:
```diff
@@ -44,6 +43,13 @@ const options = {
     // environment variable as this endpoint will return HTTP 501 if it is not set
     useSLASPrivateClient: {{answers.project.commerce.isSlasPrivate}},
 
+    // If you wish to use additional SLAS endpoints that require private clients,
```
This file is now up to date with the corresponding [ssr.js](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/vm/fix-v4-generator/packages/template-chakra-storefront/src/ssr.js) file.

vmarta commented on packages/template-chakra-storefront/src/static/manifest.json:
```diff
@@ -0,0 +1,18 @@
+{
```
Bonus: bringing back the missing manifest.json file. The .hbs version is still intact, so we just needed to bring this json file back.

vmarta commented on packages/template-chakra-storefront/src/components/_app/partials/app-seo.jsx:
```diff
@@ -19,19 +19,23 @@ import {getPathWithLocale} from '../../../utils/url'
 const AppSEO = ({appConfig, appOrigin, themeColor, site, locale, buildUrl, location}) => {
     return (
         <Seo>
-            <Helmet>
```
Bonus: fixed the warning about invalid nested Helmet. It turns out that we don't need this Helmet here, since it's already provided in the containing `Seo` component.

adamraya commented on packages/pwa-kit-create-app/assets/templates/chakra-storefront/src/components/_app-config/index.jsx.hbs:
```diff
@@ -114,22 +132,39 @@ AppConfig.propTypes = {
     locals: PropTypes.object
 }
 
+const isServerSide = typeof window === 'undefined'
```
Should we re-use the exiting `isServer` util? 

https://github.com/SalesforceCommerceCloud/pwa-kit/blob/ce815c25b3e02866199d3198ec7945b778e0fe8f/packages/template-chakra-storefront/src/utils/utils.js#L85

adamraya commented on packages/pwa-kit-create-app/assets/templates/chakra-storefront/src/components/_app-config/index.jsx.hbs:
```diff
@@ -114,22 +132,39 @@ AppConfig.propTypes = {
+        data?.queries?.forEach(updateQueryTimeStamp)
+
+        return data
+    }
```
Why are we no longer using `resetDehydratedStateTimeStamp` from commerce-sdk-react?

https://github.com/SalesforceCommerceCloud/pwa-kit/blob/63ce6a2a5e4a5912af458bde1f2b3441c703d34f/packages/commerce-sdk-react/src/utils.ts#L127

It looks like we were using that util intentionally https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2310#discussion_r1990116960

vmarta replied:
@bfeister is it all right if we switched back to calling `resetDehydratedStateTimeStamp`? The new and old implementation are very similar. They achieved the same goal but I prefer the old way of avoiding mutating the original data.

### **Old Implementation** (`resetDehydratedStateTimeStamp`)

```javascript
// Required import
import {CommerceApiProvider, resetDehydratedStateTimeStamp} from '@salesforce/commerce-sdk-react'

// Usage
const options = {
    beforeHydrate: resetDehydratedStateTimeStamp
}
```

**The utility function:**
```typescript
export const resetDehydratedStateTimeStamp = (data: DehydratedState, timestamp?: Date) => {
    const time = timestamp || Date.now()

    const updateQueryTimeStamp = (item: any) => ({
        ...item,                    // Creates new object
        state: {
            ...item.state,          // Creates new state object
            dataUpdatedAt: time
        }
    })

    return {
        mutations: data.mutations.map(updateQueryTimeStamp),  // New array
        queries: data.queries.map(updateQueryTimeStamp)       // New array
    }
}
```

### **New Implementation** (Inline Function)

```javascript
// No import needed

const options = {
    beforeHydrate: (data) => {
        const now = Date.now()

        // Helper to reset the data timestamp to time of app load.
        const updateQueryTimeStamp = ({state}) => {
            state.dataUpdatedAt = now    // Direct mutation
        }

        // Update serialized mutations and queries
        data?.mutations?.forEach(updateQueryTimeStamp)  // Modifies existing
        data?.queries?.forEach(updateQueryTimeStamp)    // Modifies existing

        return data  // Returns the same data object
    }
}
```

vmarta replied:
I've already switched it back to call `resetDehydratedStateTimeStamp`. I was calling it the "old" implementation, but actually it's the other way around.

The `develop` branch did not call resetDehydratedStateTimeStamp, but in the `next` branch it does (ever since [this PR](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2310#discussion_r1990116960), as per Adam's comment).

vmarta committed:
Merge branch 'next' into vm/fix-v4-generator

vmarta committed:
Use isServer util instead

vmarta committed:
Make sure to call resetDehydratedStateTimeStamp

vmarta committed:
Fix broken presets

All non-interactive presets should have answers.

vmarta committed:
Sync the preset values for storefront-demo

Ujwala updated them recently in the develop branch.

vmarta commented on packages/pwa-kit-create-app/assets/templates/chakra-storefront/src/components/_app-config/index.jsx.hbs:
```diff
@@ -150,21 +150,7 @@ const options = {
-
-        return data
-    }
+    beforeHydrate: resetDehydratedStateTimeStamp
```
For consistency, we're calling resetDehydratedStateTimeStamp here too.

vmarta commented on packages/pwa-kit-create-app/scripts/create-mobify-app.js:
```diff
@@ -265,22 +265,21 @@ const PRESETS = [
-                'project.dataCloud.tenantId': 'mmyw8zrxhfsg09lfmzrd1zjqmg',
-                'project.demo.enableDemoSettings': false
-            }),
+        answers: {
```
It turns out that most of the presets are broken too. But I've now fixed them.

We recently introduced `interactive: true` in the `next` branch. Those presets that are interactive will prompts the users some questions, while non-interactive presets won't ask the questions but will silently use the given answers. 

So the bug fix is making sure that non-interactive presets like this one have the `answers`.

vmarta commented on packages/pwa-kit-create-app/scripts/create-mobify-app.js:
```diff
@@ -265,22 +265,21 @@ const PRESETS = [
+        answers: {
+            'project.hybrid': false,
+            'project.name': 'demo-storefront',
+            'project.commerce.instanceUrl': 'https://production-sitegenesis-dw.demandware.net',
```
In the `develop` branch, this preset no longer uses the zzte-053 instance anymore. So I've made sure it's up to date.

vmarta committed:
Merge branch 'next' into vm/fix-v4-generator
